ParameterVector object to simplify parameter list management

dongreenberg opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

[ ] I have added the tests to cover my changes.
[ ] I have updated the documentation accordingly.
[x] I have read the CONTRIBUTING document.
-->

### Summary
A ParameterVector object to simplify instantiation, binding, and management of lists of parameters.


### Details and comments
See discussion and examples in https://github.com/Qiskit/qiskit-terra/issues/2377

kdk commented on qiskit/circuit/parametervector.py:
```diff
@@ -0,0 +1,81 @@
+    def __getitem__(self, offset):
+        return self.params[offset]
+
+    # Returns a list, not a ParameterVector
```
AFAICT, the implementation of slicing changed from python 2 -> 3. Slices are now passed directly to `__getitem__` (which works as you have it) so you should be able to remove `__getslice__`.

dongreenberg replied:
Great catch, done

kdk commented on qiskit/circuit/parametervector.py:
```diff
@@ -0,0 +1,81 @@
+    def __iter__(self):
+        return iter(self.params)
+
+    def __next__(self):
```
Since `__iter__` returns an interator over the `self.param` list, I don't think this will be called.

dongreenberg replied:
Good call, fixed

kdk commented on qiskit/circuit/parametervector.py:
```diff
@@ -0,0 +1,81 @@
+    def __str__(self):
+        return '{}, {}'.format(self.name, [str(item) for item in self.params])
+
+    def __copy__(self):
```
If `ParameterVector` is mutable, would we want this to be a normal copy? Say if I have a vector, copy it, and then extend the copy, do I want those changes to also appear in the original? (The way `Parameter` implements copy is sort of a hack that I'd like to revisit, but I don't think we need that behavior here.)

dongreenberg replied:
I think the standard is for the array to be copied properly, but not the elements in the array - meaning if you added elements to your copy it wouldn't be reflected in the original, but if you changed an element object it would. Should we do that? If so, should I get rid of __deepcopy__?

kdk replied:
Yes, I think that's the behavior we want. With the `ParameterVector` immutable, this should be a non-issue, but unless there's a use case for the non-standard copy, I think we can remove `__copy__` and `__deepcopy__`.

dongreenberg replied:
Makes sense, removing

kdk commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -818,14 +819,21 @@ def bind_parameters(self, value_dict):
+            if isinstance(param, Parameter):
+                unrolled_value_dict[param] = value
+            if isinstance(param, ParameterVector):
+                unrolled_value_dict.update(zip(param, value))
```
Can you add a check that `param` and `value` are the same length? (`zip` will stop the end of whichever is shorter)

dongreenberg replied:
Done

dongreenberg commented:
Added support for passing vector remapping into to_instruction
Instruction-based construction added to tests
Turned @kdk 's comments:
Changed autonames of Parameter elements to 'vecname[index]'
Fixed iterator interface mistakes
Made ParameterVector immutable

Tests pass

1ucian0 commented on test/python/circuit/test_parameters.py:
```diff
@@ -170,3 +172,70 @@ def test_parameter_name_conflicts_raises(self):
+            for i, q in enumerate(qc.qubits[:-1]):
+                qc.cx(qc.qubits[i], qc.qubits[i+1])
+            qc.barrier()
+        theta_vals = np.random.random(len(theta)) * np.pi
```
Can you fix this value? Test should be deterministic #205 :)

dongreenberg replied:
Fixed

ajavadia commented on test/python/circuit/test_parameters.py:
```diff
@@ -170,3 +170,73 @@ def test_parameter_name_conflicts_raises(self):
+        depth = 4
+        ryrz = QuantumCircuit(qubits, name='ryrz')
+        theta = ParameterVector('θ0', length=len(ryrz.qubits) * 2)
+        theta_iter = iter(theta)
```
I think you can make the ParameterVector object itself iterable (just define `__iter__` and `__next__`). That way you don't have to explicitly convert to iterable here, and you could do things like `for single_theta in theta: ...`. This seems more aligned with the goal of making this class easy to work with and not having to rely on Python loops, iterations, etc.

dongreenberg replied:
The only reason I have the `iter` call here is because I call `next` twice inside the loop, and for some reason there's no clean way that I know of to iterate in a for loop in buckets of n items (e.g. for each step give me four items). The iteration you want works here too, because when you call `for theta_n in theta` it will call `iter(theta)`, get the list iterator, and then each theta will be the `Parameter` objects in the list. I originally had `next()` implemented and it was just calling `next()` on a list iterator which was acquired and saved inside `ParameterVector.__iter__()` but realized we don't need it and it's a little messy. Does that make sense?

ajavadia replied:
I just pushed [61b1067](https://github.com/Qiskit/qiskit-terra/pull/2379/commits/61b1067a5833d64243ab40bf666ca3e0637d0663) to your branch, which I think does what I'm saying here. Basically now you can call `iter(theta_vector)` -- it behaves like a normal iterator. And the explicit conversion to `iter` is not needed anymore.

kdk replied:
The problem with making `ParameterVector` the an iterator is that iterators are inherently single-use.

```
>>> v = ParameterVector('v', 3)
>>> next(v)
Parameter(v[0])
>>> next(v)
Parameter(v[1])
>>> next(v)
Parameter(v[2])
>>> next(v)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/parametervector.py", line 49, in __next__
    raise StopIteration
StopIteration
```
Now there's no way to go back and reset the iterator over `v`. For example, `List`s are not interators, but you can generate an interator from a `List`.

```
>>> b = [1,2,3]
>>> next(b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'list' object is not an iterator
>>> i = iter(b)
>>> next(i)
1
>>> next(i)
2
>>> i2 = iter(b)
>>> next(i2)
1
>>> next(i)
3
```

Prior to https://github.com/Qiskit/qiskit-terra/pull/2379/commits/61b1067a5833d64243ab40bf666ca3e0637d0663 , it was still possible to write `for param in vector: ...` since `ParameterVector.__iter__` would return an iterator over `self._params`, which would have it's own `__iter__` and `__next__`.

The reason the RyRz test explicitly calls `iter(ParameterVector)` is because its implemented by pulling off parameters one by one in sequence, rather than by drawing groups of parameters at once.

e.g. If I have a a group of three gates I'd like to repeat five times, I could write either:
```
v = ParameterVector('v', 15)
v_iter = iter(v)
for group_index in range(5):
  qc.rx(next(v_iter), 0)
  qc.ry(next(v_iter), 0)
  qc.rz(next(v_iter), 0)
```
or which, to me is the python equivalent of something like:
```
v = ParameterVector('v', 15)
i = 0
while i < len(v):
  qc.rx(v[i++], 0)
  qc.ry(v[i++], 0)
  qc.rz(v[i++], 0)
```

@dongreenberg is using `iter` as a means of keeping track of position within the `ParameterVetor`, not because it didn't natively support iteration.

ajavadia replied:
Depends on whether you want this object to behave purely like a list, or to also be able to use it as an iterator. If you think that being able to do the following is useful, then `__next__` does that.
```python
In [2]: v = ParameterVector('v', 3)

In [3]: next(v)
Out[3]: Parameter(v[0])
```
Note that this does not preclude the previous behavior, i.e. you can still do:
```python
In [4]: i = iter(v)

In [5]: next(i)
Out[5]: Parameter(v[0])

In [6]: next(i)
Out[6]: Parameter(v[1])

In [7]: i2 = iter(v)

In [8]: next(i2)
Out[8]: Parameter(v[0])

In [9]: next(i)
Out[9]: Parameter(v[2])
```
But I'm fine reverting the commit, it was mostly to simplify those tests.

dongreenberg replied:
Oh oh oh I didn't catch that. I like that, but leave it up to kdk because I'm a shorthand liberal.

kdk replied:
I like the conciseness, but to me, it pushes `ParameterVector` too far from convention. It's a list-like thing, and you can use it in most of the standard like-like ways. (e.g. "If you want it as an interator, use `iter`.") Also, that it can only be single use as an iterator will only ever confuse someone who tries to use it that way.

(As an aside, it looks like an iterator is [expected](https://docs.python.org/3/library/stdtypes.html#iterator.__iter__) to return `self` on it's `__iter__` method, though I can't off the top of my head see what this would break.)

ajavadia replied:
Ok I'm fine with having this object as just a list. I just reverted the commit.

dongreenberg commented:
Bump?

ajavadia commented:
@kdk @dongreenberg I made a small change, but overall I'm good now. Starting to really like this actually
<img width="710" alt="image" src="https://user-images.githubusercontent.com/8622381/57886188-a37e3f00-77fa-11e9-8c08-9710a5b29ef7.png">

dongreenberg commented:
Thanks, @kdk and @ajavadia for your help and contributions!
